xpart/particles/particles.py

```diff
@@ -157,109 +159,346 @@ class Particles(xo.HybridClass):
+            _gamma0 = gamma0
+        elif beta0 is not None:
+            _gamma0 = 1 / self._sqrt(1 - beta0 ** 2)
+            _energy0 = self.mass0 * _gamma0
+            _p0c = _energy0 * beta0
+            _beta0 = beta0
+        else:
+            raise RuntimeError('This statement is unreachable.')
+
+        self._assert_values_consistent(energy0, self.mass0 * _gamma0, mask)
+        self._setattr_if_consistent('_p0c', p0c, _p0c, mask)
+        self._setattr_if_consistent('_gamma0', gamma0, _gamma0, mask)
+        self._setattr_if_consistent('_beta0', beta0, _beta0, mask)
+
+    def _update_energy_deviations(self, delta=None, ptau=None, pzeta=None,
+                                  _rpp=None, _rvv=None, mask=None):
+        if all(ff is None for ff in (delta, ptau, pzeta)):
+            if _rpp is not None or _rvv is not None:
+                raise ValueError('Setting `delta` and `ptau` by only giving '
+                                 '`_rpp` and `_rvv` is not supported.')
+            self._delta = 0.0
+            delta = self._delta  # Cupy complains if we later assign LinkedArray
+
+        beta0 = self._beta0
+        if delta is not None:
+            _delta = delta
+            _ptau = self._sqrt(_delta**2 + 2 * _delta + 1 / beta0**2) - 1 / beta0
+            _pzeta = _ptau / beta0
+        elif ptau is not None:
+            _ptau = ptau
+            _delta = self._sqrt(_ptau ** 2 + 2 * _ptau / beta0 + 1) - 1
+            _pzeta = _ptau / beta0
+        elif pzeta is not None:
+            _pzeta = pzeta
+            _ptau = _pzeta * beta0
+            _delta = self._sqrt(_ptau ** 2 + 2 * _ptau / beta0 + 1) - 1
+        else:
+            raise RuntimeError('This statement is unreachable.')
+
+        self._assert_values_consistent(pzeta, _pzeta, mask)
+        self._setattr_if_consistent('_delta', delta, _delta, mask)
+        self._setattr_if_consistent('_ptau', ptau, _ptau, mask)
+
+        delta = self._delta  # Cupy complains if we later assign LinkedArray
+        delta_beta0 = delta * beta0
+        ptau_beta0 = self._sqrt(delta_beta0 ** 2 + 2 * delta_beta0 * beta0 + 1) - 1
+        __rvv = (1 + delta) / (1 + ptau_beta0)
+        __rpp = 1 / (1 + delta)
+        self._setattr_if_consistent('_rpp', _rpp, __rpp, mask)
+        self._setattr_if_consistent('_rvv', _rvv, __rvv, mask)
+
+    def _update_zeta(self, zeta=None, tau=None, mask=None):
+        if zeta is None and tau is None:
+            self.zeta = 0.0
+            zeta = self.zeta
+
+        beta0 = self._beta0
+
+        if zeta is not None:
+            _zeta = zeta
+            _tau = zeta / beta0
+        elif tau is not None:
+            _tau = tau
+            _zeta = beta0 * _tau
+        else:
+            raise RuntimeError('This statement is unreachable.')
+
+        self._assert_values_consistent(tau, _tau, mask)
+        self._setattr_if_consistent('zeta', zeta, _zeta, mask)
+
+    def _update_chi_charge_ratio(self, chi=None, charge_ratio=None,
```

Comment:
Check consistency with the physics guide:
https://github.com/xsuite/xsuite/blob/main/docs/physics_manual/physics_man.pdf